File: testcasegenerator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX_FILES (int)1e6
#define MAX_NAME_LENGTH 6
#define START_YEAR 2023
#define MAX_ID (int)1e6

// Generates a unique ID not in the generated_ids array
int generate_unique_id(int *generated_ids, int count) {
    int id;
    int unique;
    do {
        unique = 1;
        id = rand() % MAX_ID + 1;
        for (int i = 0; i < count; i++) {
            if (generated_ids[i] == id) {
                unique = 0;
                break;
            }
        }
    } while (!unique);
    return id;
}

// Generates a random ISO 8601 timestamp within the year START_YEAR
void generate_random_timestamp(char *timestamp) {
    struct tm time;
    time.tm_year = START_YEAR - 1900;
    time.tm_mon = rand() % 12;
    time.tm_mday = rand() % 28 + 1;
    time.tm_hour = rand() % 24;
    time.tm_min = rand() % 60;
    time.tm_sec = rand() % 60;

    strftime(timestamp, 20, "%Y-%m-%dT%H:%M:%S", &time);
}

// Generates a random filename of length len
void generate_random_filename(char *filename, int len) {
    const char charset[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    for (int i = 0; i < len - 1; i++) {
        filename[i] = charset[rand() % (sizeof(charset) - 1)];
    }
    filename[len - 1] = '\0';  // Null-terminate the string
}
// Main test case generator function
void generate_test_case(int num_files) {
    const char *sorting_criteria[] = {"Name", "ID", "Timestamp"};
    int ids[MAX_FILES] = {0};  // Array to track unique IDs
    FILE *file = fopen("test_case.txt", "w");

    if (file == NULL) {
        perror("Unable to create test case file");
        exit(EXIT_FAILURE);
    }

    fprintf(file, "%d\n", num_files);
    if(num_files>1e6){
        
    }
    for (int i = 0; i < num_files; i++) {
        char file_name[MAX_NAME_LENGTH];
        int file_id = generate_unique_id(ids, i);
        ids[i] = file_id;
        char timestamp[20];

        
        int len = rand() % (MAX_NAME_LENGTH) + 3;
        generate_random_filename(file_name, len);
        generate_random_timestamp(timestamp);

        fprintf(file, "%s.txt %d %s\n", file_name, file_id, timestamp);
    }

    // Choose a random sorting criterion
    const char *criterion = sorting_criteria[rand() % 3];
    fprintf(file, "%s\n", criterion);

    fclose(file);
    printf("Test case generated in 'test_case.txt'\n");
}

int main() {
    srand((unsigned int)time(NULL));  // Seed random number generator

    int num_files;
    printf("Enter the number of files to generate: ");
    scanf("%d", &num_files);
    if (num_files > MAX_FILES) {
        fprintf(stderr, "Max number of files is %d\n", MAX_FILES);
        return EXIT_FAILURE;
    }

    generate_test_case(num_files);
    return EXIT_SUCCESS;
}

File: scripts/testcasegenerator.py
import random
import string
import time
import concurrent.futures
from datetime import datetime

MAX_FILES = int(1e8)
MAX_NAME_LENGTH = int(5)
MAX_ID = int(1e5+1)

# Generates a random START_YEAR between 2000 and current year
def generate_random_year():
    current_year = datetime.now().year
    return random.randint(2022, current_year)

# Generates a unique ID not in the generated IDs set
def generate_unique_id(generated_ids):
    while True:
        id = random.randint(1, MAX_ID)
        if id not in generated_ids:
            generated_ids.add(id)
            return id

# Generates a random ISO 8601 timestamp within a random year
def generate_random_timestamp():
    year = generate_random_year()  # Random year now
    month = random.randint(1, 12)
    day = random.randint(1, 28)  # Keeping it simple for month-day validation
    hour = random.randint(0, 23)
    minute = random.randint(0, 59)
    second = random.randint(0, 59)
    return f"{year}-{month:02d}-{day:02d}T{hour:02d}:{minute:02d}:{second:02d}"


# Generates a random filename of length `length` using only alphabetic characters
def generate_random_filename(length):
    charset = string.ascii_letters  # Use only alphabetic characters
    return ''.join(random.choice(charset) for _ in range(length))


# Function to generate test case for a chunk of files
def generate_file_chunk(start_index, end_index, generated_ids):
    files_chunk = []

    for _ in range(start_index, end_index):
        file_name_length = random.randint(3, MAX_NAME_LENGTH)
        file_name = generate_random_filename(file_name_length)
        file_id = generate_unique_id(generated_ids)
        timestamp = generate_random_timestamp()
        files_chunk.append(f"{file_name}.txt {file_id} {timestamp}")
    
    return files_chunk

# Main test case generator function using multithreading
def generate_test_case(num_files):
    # Track unique IDs in a set
    generated_ids = set()
    
    # Split the workload into chunks for multithreading
    chunk_size = num_files // 8  # Divide work into 8 parts for multithreading, adjust as needed
    chunks = [(i * chunk_size, (i + 1) * chunk_size) for i in range(8)]
    chunks[-1] = (chunks[-1][0], num_files)  # Make sure the last chunk includes all remaining files
    
    # Choose a random sorting criterion
    sorting_criteria = ["Name", "ID", "Timestamp"]
    criterion = random.choice(sorting_criteria)
    criterion = "Name"  # Hardcoded for now
    
    with open("test_case_count_final.txt", "w") as file:
        file.write(f"{num_files}\n")
        
        # Use ThreadPoolExecutor to process file chunks in parallel
        with concurrent.futures.ThreadPoolExecutor() as executor:
            futures = []
            for start_idx, end_idx in chunks:
                futures.append(executor.submit(generate_file_chunk, start_idx, end_idx, generated_ids))

            # Collect results from each chunk and write to file
            for future in concurrent.futures.as_completed(futures):
                chunk = future.result()
                file.writelines([line + '\n' for line in chunk])  # Write chunk to file
        
        # Write the sorting criterion at the end of the file
        file.write(f"{criterion}\n")

    print("Test case generated in 'test_case_count_final.txt'")

import sys
def main():
    random.seed(int(time.time()))  # Seed the random number generator

    # num_files = int(input("Enter the number of files to generate: "))
    # num_files = random.randint(1, MAX_FILES)
    num_files = int(sys.argv[1])
    
    if num_files > MAX_FILES:
        print(f"Max number of files is {MAX_FILES}")
        return

    generate_test_case(num_files)

if __name__ == "__main__":
    main()
